### ISSUE TITLE
invalid-assignment when setting `warnings.showwarning`

### ISSUE THREAD

**Issue by fwdekker-pc** [astral-sh/ty#3962]
### Summary

If you set `warnings.showwarning`, ty will error with `invalid-assignment`, even if the type of the assignment is exactly correct.

```python
import warnings
from typing import TextIO


def foo(
    message: Warning | str,
    category: type[Warning],
    filename: str,
    lineno: int,
    file: TextIO | None = None,
    line: str | None = None,
) -> None:
    print("Oops!")


warnings.showwarning = foo
```

gives

```
error[invalid-assignment]: Object of type `def foo(message: Warning | str, category: type[Warning], filename: str, lineno: int, file: TextIO | None = None, line: str | None = None) -> None` is not assignable to attribute `showwarning` of type `def showwarning(message: Warning | str, category: type[Warning], filename: str, lineno: int, file: TextIO | None = None, line: str | None = None) -> None`
  --> src\blablabla\__main__.py:16:1
   |
16 | warnings.showwarning = foo
   | ^^^^^^^^^^^^^^^^^^^^
   |

Found 1 diagnostic
error: recipe `check` failed on line 65 with exit code 1
```

Compare the actual and desired types:
```
def foo        (message: Warning | str, category: type[Warning], filename: str, lineno: int, file: TextIO | None = None, line: str | None = None) -> None
def showwarning(message: Warning | str, category: type[Warning], filename: str, lineno: int, file: TextIO | None = None, line: str | None = None) -> None
```

Renaming `foo` to `showwarning` does not resolve the issue.

Using Python 3.14.6 on Windows 11 25H2.

### Version

ty 0.0.58 (17d3f6db5 2026-07-09)

**Comment by carljm** [comment]
Thanks for the report!

Since `showwarnings` is not an annotated variable, we infer a precise type for it, which is the singleton type of "the exact function defined at that location" rather than a more general callable type of all functions sharing a compatible signature. So in general we do not permit monkey-patching module-level functions.

But in this case, monkey-patching `warnings.showwarnings` is a [documented customization hook](https://docs.python.org/3/library/warnings.html). So we should support it.

We could generally infer a broader callable type for all module-level functions -- this would require ensuring that we track all the specific-function details in that callable type, for LSP purposes.

Or we could leave inference the same but special-case allowing monkeypatching with a function with compatible signature, technically making the function-literal type inaccurate, but in practice it's probably fine? I think this is how we currently handle monkeypatching methods on classes.

In a sense I think the first is the better option, since if a FunctionLiteral is no longer really reliably a singleton type, probably we should stop having that type at all and just use Callable types that track origin information. But the second is a much less invasive change.
